src/python/pants/backend/python/target_types.py

```diff
@@ -465,29 +463,31 @@ def format_invalid_requirement_string_error(
     )
 
 
-class PythonRequirementsField(PrimitiveField):
+class PythonRequirementsField(PrimitiveField[Tuple[Requirement, ...]]):
     """A sequence of pip-style requirement strings, e.g. ['foo==1.8', 'bar<=3 ;
     python_version<'3']."""
 
     alias = "requirements"
     required = True
-    value: Tuple[Requirement, ...]
 
     @classmethod
-    def compute_value(
-        cls, raw_value: Optional[Iterable[str]], *, address: Address
+    def sanitize_raw_value(
+        cls, raw_value: Optional[Union[str, PythonRequirement, Iterable[str]]], *, address: Address
     ) -> Tuple[Requirement, ...]:
-        value = super().compute_value(raw_value, address=address)
-        invalid_type_error = InvalidFieldTypeException(
-            address,
-            cls.alias,
-            value,
-            expected_type="an iterable of pip-style requirement strings (e.g. a list)",
-        )
+        value = super().sanitize_raw_value(raw_value, address=address)
+
+        def invalid_type_error():
+            raise InvalidFieldTypeException(
+                address,
+                cls.alias,
+                value,
+                expected_type="an iterable of pip-style requirement strings (e.g. a list)",
+            )
+
         if isinstance(value, (str, PythonRequirement)) or not isinstance(
             value, collections.abc.Iterable
         ):
-            raise invalid_type_error
+            raise invalid_type_error()
```

Comment:
You're now calling `raise` twice, here and in the function. 

I'm not sure this is much of an improvement, tbh. I intentionally had it this way to avoid the function having the side effect of calling `raise`. That way, it's very explicit where `raise` is being called.

Reply:
Oh oops, I guess I meant `return`. I also know from checking [the `cpython` crate docs](https://docs.rs/cpython/0.5.1/cpython/struct.PyErr.html) that exceptions are actually constructed lazily to some extent? I was concerned that creating an exception unconditionally even in the success case could be slow. I will change the function to `return` the value -- let me know your thoughts on that.

Reply:
There's nothing to return, as `raise` causes the return type to be `NoReturn`. So, either:

1) Restore to what we had before.
2) Make your change, and at the call sites say `invalid_type_error`. Perhaps rename the function to `raise_invalid_type_error` for clarity.

I'd recommend #1. The code worked before, and this is a subjective style preference between #1 and #2. In situations like that, I think we typically avoid making changes to reduce churn.

Reply:
We can return the exception to raise and annotate with `-> Exception:`, I was thinking.

Reply:
Oh, sure, I'm comfortable with that! I like that much more than putting `raise` in the function itself. I still tbh think it's overkill to use a lazy constructor, but not a big deal.

Main thing I care about is using `raise` at the place where the error happens.